src/sentry/scripts/similarity/index.lua

```diff
@@ -0,0 +1,618 @@
+--[[
+
+Implements an index that can be used to efficiently search for items that
+share similar characteristics.
+
+This implementation is based on MinHash (which is used quickly identify
+similar items and estimate the Jaccard similarity of their characteristic
+sets) but this implementation extends the typical design to add the ability
+to record items by an arbitrary key. This allows querying for similar
+groups that contain many different characteristic sets.
+
+This is modeled as two data structures:
+
+- A bucket frequency hash, which maintains a count of what buckets
+    have been recorded -- and how often -- in a ``(band, key)`` pair. This
+    data can be used to identify what buckets a key is a member of, and also
+    used to identify the degree of bucket similarity when comparing with data
+    associated with another key.
+- A bucket membership set, which maintains a record of what keys have been
+    record in a ``(band, bucket)`` pair. This data can be used to identify
+    what other keys may be similar to the lookup key (but not the degree of
+    similarity.)
+
+]]--
+local function range(start, stop)
```

Comment:
I have the same feeling about `imap`, `izip`, and `ifilter` as well to be honest, I might make this change if it's a quick one but am not dead set on changing it before merging this. (Also, for posterity, these are prefixed with `i` not to represent "iterator" like in Python, but to represent that they use `ipairs` and not `pairs`.)